commit f4dc83d8926509d0958ec21fcdbc2e7df3d32ce2
Author: John Hawthorn <john@hawthorn.email>
Date:   Thu Oct 10 20:32:00 2024 -0700

    Avoid backtracking in Token#raw_params
    
    Thanks to scyoon for the patch
    
    [CVE-2024-47887]

diff --git a/actionpack/lib/action_controller/metal/http_authentication.rb b/actionpack/lib/action_controller/metal/http_authentication.rb
index 017e7a3a2f..bb60ffdeba 100644
--- a/actionpack/lib/action_controller/metal/http_authentication.rb
+++ b/actionpack/lib/action_controller/metal/http_authentication.rb
@@ -511,31 +511,28 @@ def params_array_from(raw_params)
       # This removes the `"` characters wrapping the value.
       def rewrite_param_values(array_params)
         array_params.each { |param| (param[1] || +"").gsub! %r/^"|"$/, "" }
       end
 
-      WHITESPACED_AUTHN_PAIR_DELIMITERS = /\s*#{AUTHN_PAIR_DELIMITERS}\s*/
-      private_constant :WHITESPACED_AUTHN_PAIR_DELIMITERS
-
       # This method takes an authorization body and splits up the key-value pairs by
       # the standardized `:`, `;`, or `\t` delimiters defined in
       # `AUTHN_PAIR_DELIMITERS`.
       def raw_params(auth)
-        _raw_params = auth.sub(TOKEN_REGEX, "").split(WHITESPACED_AUTHN_PAIR_DELIMITERS)
+        _raw_params = auth.sub(TOKEN_REGEX, "").split(AUTHN_PAIR_DELIMITERS).map(&:strip)
         _raw_params.reject!(&:empty?)
 
         if !_raw_params.first&.start_with?(TOKEN_KEY)
           _raw_params[0] = "#{TOKEN_KEY}#{_raw_params.first}"
         end
 
         _raw_params
       end
 
       # Encodes the given token and options into an Authorization header value.
       #
       # Returns String.
       #
       # #### Parameters
       #
       # *   `token` - String token.
       # *   `options` - Optional Hash of the options.
       #
